File: Sample Code/JavaScript/Snake - ECS -n- OO/Snake - ECS/scripts/game-model.js
function GameModel() {
    'use strict';
    const GRID_SIZE = 50;
    const OBSTACLE_COUNT = 15;
    const MOVE_INTERVAL = 150;

    let entities = {};  // key is 'id', value is an Entity

    // --------------------------------------------------------------
    //
    // Defining the game border as entities that have position, collision,
    // and visual components.
    //
    // --------------------------------------------------------------
    function initializeBorder() {
        let border = {};

        function createBorderEntity(x, y) {
            let border = Entity.createEntity();
            border.addComponent(MyGame.components.Appearance({ fill: { r: 255, g: 0, b: 0 }, stroke: 'rgb(0, 0, 0)' }));
            border.addComponent(MyGame.components.Position({ segments: [{ x: x, y: y }] }));
            border.addComponent(MyGame.components.Collision());

            return border;
        }

        for (let position = 0; position < GRID_SIZE; position++) {
            // Left border
            let left = createBorderEntity(0, position);
            border[left.id] = left;
            // Right border
            let right = createBorderEntity(GRID_SIZE - 1, position);
            border[right.id] = right;
            // Top border
            let top = createBorderEntity(position, 0);
            border[top.id] = top;
            // bottom border
            let bottom = createBorderEntity(position, GRID_SIZE - 1);
            border[bottom.id] = bottom;
        }

        return border;
    }

    // --------------------------------------------------------------
    //
    // Defining each of the obstacles as entities that have position,
    // collision, and visual components.
    //
    // --------------------------------------------------------------
    function initializeObstacles() {
        let obstacles = {};

        function createObstacleEntity(x, y) {
            let obstacle = Entity.createEntity();
            obstacle.addComponent(MyGame.components.Appearance({ fill: {r: 0, g: 255, b: 0 }, stroke: 'rgb(0, 0, 0)' }));
            obstacle.addComponent(MyGame.components.Position({ segments: [{ x: x, y: y }] }));
            obstacle.addComponent(MyGame.components.Collision());

            return obstacle;
        }

        let remaining = OBSTACLE_COUNT;
        while (remaining > 0) {
            let x = Random.nextRange(1, GRID_SIZE - 1);
            let y = Random.nextRange(1, GRID_SIZE - 1);
            let proposed = createObstacleEntity(x, y);
            if (!MyGame.systems.collision.collidesWithAny(proposed, obstacles)) {
                obstacles[proposed.id] = proposed;
                remaining--;
            }
        }

        return obstacles;
    }

    // --------------------------------------------------------------
    //
    // Defining the food as an entity that has position, collision, 
    // and visual components.
    //
    // --------------------------------------------------------------
    function createFood() {
        let food = null;

        function createFoodEntity(x, y) {
            let food = Entity.createEntity();
            food.addComponent(MyGame.components.Appearance({ fill: {r: 255, g: 128, b: 0 }, stroke: 'rgb(0, 0, 0)' }));
            food.addComponent(MyGame.components.Position({ segments: [{ x: x, y: y }] }));
            food.addComponent(MyGame.components.Collision());
            food.addComponent(MyGame.components.Food());

            return food;
        }

        let done = false;
        while (!done) {
            let x = Random.nextRange(1, GRID_SIZE - 1);
            let y = Random.nextRange(1, GRID_SIZE - 1);
            //
            // Create a proposed food entity at this location and see if it collides with anything
            let proposed = createFoodEntity(x, y);
            if (!MyGame.systems.collision.collidesWithAny(proposed, entities)) {
                food = proposed;
                done = true;
            }
        }

        return food;
    }

    // --------------------------------------------------------------
    //
    // Defining the snake as an entity that has position, direction,
    // collision, visual, and input components.
    //
    // --------------------------------------------------------------
    function initializeSnake() {
        let snake = null;

        function createSnakeEntity(x, y) {
            let snake = Entity.createEntity();
            snake.addComponent(MyGame.components.Appearance({ fill: { r: 255, g: 255, b: 255 }, stroke: 'rgb(0, 0, 0)' }));
            snake.addComponent(MyGame.components.Position({ segments: [{ x: x, y: y }] }));
            snake.addComponent(MyGame.components.Collision());
            snake.addComponent(MyGame.components.Movable({ facing: MyGame.constants.Direction.Stopped, moveInterval: MOVE_INTERVAL }));
            let inputSpecification = { keys: {
                'ArrowLeft': MyGame.constants.Direction.Left,
                'ArrowRight': MyGame.constants.Direction.Right,
                'ArrowUp': MyGame.constants.Direction.Up,
                'ArrowDown': MyGame.constants.Direction.Down
            }};
            snake.addComponent(MyGame.components.KeyboardControlled(inputSpecification));

            return snake;
        }

        let done = false;
        while (!done) {
            let x = Random.nextRange(1, GRID_SIZE - 1);
            let y = Random.nextRange(1, GRID_SIZE - 1);
            //
            // Create a proposed snake entity at this location and see if it collides with anything
            let proposed = createSnakeEntity(x, y);
            if (!MyGame.systems.collision.collidesWithAny(proposed, entities)) {
                snake = proposed;
                done = true;
            }
        }

        return snake;
    }

    // --------------------------------------------------------------
    //
    // Interface that allows systems to report events back to the overall
    // game model for processing.
    //
    // --------------------------------------------------------------
    function reportEvent(info) {
        switch (info.type) {
            case 'consume-food':
                delete entities[info.entity.id];
                let food = createFood();
                entities[food.id] = food;
                break;
            case 'hit-something':
                break;
        }
    }

    // --------------------------------------------------------------
    //
    // Helper method used to merge the properites of the 'source' object
    // into the 'dest' object.
    //
    // --------------------------------------------------------------
    function mergeObjects(dest, source) {
        for (let key in source) {
            dest[key] = source[key];
        }
    }

    // --------------------------------------------------------------
    //
    // One time initialization to get the game model prepared for playing.
    //
    // --------------------------------------------------------------
    function initialize() {
        console.log('initializing borders...');
        mergeObjects(entities, initializeBorder());

        console.log('initializing obstacles...');
        mergeObjects(entities, initializeObstacles());

        console.log('initialzing snake starting position...');
        let snake = initializeSnake();
        entities[snake.id] = snake;

        console.log('initialzing first food location...');
        let food = createFood();
        entities[food.id] = food;
    }

    // --------------------------------------------------------------
    //
    // Public interface exposed to the game loop that tells the game
    // model to perform an update.
    //
    // --------------------------------------------------------------
    function update(elapsedTime) {
        MyGame.systems.keyboardInput.update(elapsedTime, entities);
        MyGame.systems.movement.update(elapsedTime, entities);
        MyGame.systems.collision.update(elapsedTime, entities, reportEvent);
        MyGame.systems.render.update(elapsedTime, entities);
    }

    initialize();

    let api = {
        update: update
    };

    return api;
}

File: Sample Code/MonoGame/ECS - Snake Game/ECS - Snake Game/SnakeGame/GameModel.cs
using Entities;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Graphics;
using System.Collections.Generic;

namespace CS5410
{
    class GameModel
    {
        private const int GRID_SIZE = 50;
        private const int OBSTACLE_COUNT = 15;
        private readonly int WINDOW_WIDTH;
        private readonly int WINDOW_HEIGHT;

        private List<Entity> m_removeThese = new List<Entity>();
        private List<Entity> m_addThese = new List<Entity>();

        private Systems.Renderer m_sysRenderer;
        private Systems.Collision m_sysCollision;
        private Systems.Movement m_sysMovement;
        private Systems.KeyboardInput m_sysKeyboardInput;

        public GameModel(int width, int height)
            {
            this.WINDOW_WIDTH = width;
            this.WINDOW_HEIGHT = height;
        }

        public void Initialize(ContentManager content, SpriteBatch spriteBatch)
        {
            var texSquare = content.Load<Texture2D>("Images/square");

            m_sysRenderer = new Systems.Renderer(spriteBatch, texSquare, WINDOW_WIDTH, WINDOW_HEIGHT, GRID_SIZE);
            m_sysCollision = new Systems.Collision((entity) =>
            {
                // Remove the existing food pill
                m_removeThese.Add(entity);
                // Need another food pill
                m_addThese.Add(createFood(texSquare));
            });
            m_sysMovement = new Systems.Movement();
            m_sysKeyboardInput = new Systems.KeyboardInput();

            initializeBorder(texSquare);
            initializeObstacles(texSquare);
            initializeSnake(texSquare);
            AddEntity(createFood(texSquare));
        }

        public void Update(GameTime gameTime)
        {
            m_sysKeyboardInput.Update(gameTime);
            m_sysMovement.Update(gameTime);
            m_sysCollision.Update(gameTime);

            foreach (var entity in m_removeThese)
            {
                RemoveEntity(entity);
            }
            m_removeThese.Clear();

            foreach (var entity in m_addThese)
            {
                AddEntity(entity);
            }
            m_addThese.Clear();
        }

        public void Draw(GameTime gameTime)
        {
            m_sysRenderer.Update(gameTime);
        }

        private void AddEntity(Entity entity)
        {
            m_sysKeyboardInput.Add(entity);
            m_sysMovement.Add(entity);
            m_sysCollision.Add(entity);
            m_sysRenderer.Add(entity);
        }

        private void RemoveEntity(Entity entity)
        {
            m_sysKeyboardInput.Remove(entity.Id);
            m_sysMovement.Remove(entity.Id);
            m_sysCollision.Remove(entity.Id);
            m_sysRenderer.Remove(entity.Id);
        }

        private void initializeBorder(Texture2D square)
        {
            for (int position = 0; position < GRID_SIZE; position++)
            {
                var left = BorderBlock.create(square, 0, position);
                AddEntity(left);

                var right = BorderBlock.create(square, GRID_SIZE - 1, position);
                AddEntity(right);

                var top = BorderBlock.create(square, position, 0);
                AddEntity(top);

                var bottom = BorderBlock.create(square, position, GRID_SIZE - 1);
                AddEntity(bottom);
            }
        }

        private void initializeObstacles(Texture2D square)
        {
            MyRandom rnd = new MyRandom();
            int remaining = OBSTACLE_COUNT;

            while (remaining > 0)
            {
                int x = (int)rnd.nextRange(1, GRID_SIZE - 1);
                int y = (int)rnd.nextRange(1, GRID_SIZE - 1);
                var proposed = Obstacle.create(square, x, y);
                if (!m_sysCollision.collidesWithAny(proposed))
                {
                    AddEntity(proposed);
                    remaining--;
                }
            }
        }

        private void initializeSnake(Texture2D square)
        {
            MyRandom rnd = new MyRandom();
            bool done = false;

            while (!done)
            {
                int x = (int)rnd.nextRange(1, GRID_SIZE - 1);
                int y = (int)rnd.nextRange(1, GRID_SIZE - 1);
                var proposed = SnakeSegment.create(square, x, y);
                if (!m_sysCollision.collidesWithAny(proposed))
                {
                    AddEntity(proposed);
                    done = true;
                }
            }
        }

        private Entity createFood(Texture2D square)
        {
            MyRandom rnd = new MyRandom();
            bool done = false;

            while (!done)
            {
                int x = (int)rnd.nextRange(1, GRID_SIZE - 1);
                int y = (int)rnd.nextRange(1, GRID_SIZE - 1);
                var proposed = Food.create(square, x, y);
                if (!m_sysCollision.collidesWithAny(proposed))
                {
                    return proposed;
                }
            }

            return null;
        }
    }
}
